### ISSUE TITLE
[RFC] Judge Framework and Online DPO

### ISSUE THREAD

**Issue by sam-pi** [pytorch/torchtune#2413]
TRL has a concept of multiple different judges which can be used in various online RLHF type methods, see the [TRL Judges Doc](https://github.com/huggingface/trl/blob/main/docs/source/judges.md). As a starting point, we could implement just a pairwise judge that could be useful with techniques like Online DPO similar to [TRL Online DPO Trainer](https://huggingface.co/docs/trl/en/online_dpo_trainer).

The basic idea for simple online DPO is to input a prompt dataset, generate two candidate outputs from the policy model, judge these to get chosen and rejected using a specified judge, and then output a preference dataset format.

Here is a first rough cut of a PairwiseJudge that has been working for me in early online DPO experiments. I've included an example below of a pairwise judge based on response length. Please let me know if you have feedback on this approach or if there could be other more useful abstractions to make this something torchtune would want to integrate.

## Pairwise Judge Proposal
```python
class PairwiseJudge:
    """
    Base class for pairwise judges.

    Provides a common interface for executing judgment functions on pairwise completions.
    """

    def __init__(self, rank: int = None):
        self.rank = rank

    def judge(self, prompt: str, candidates: List[str]) -> int:
        """
        Abstract method to be implemented by subclasses.

        Args:
            prompt (str): The input prompt.
            candidates (List[str]): List of candidate responses.

        Returns:
            int: Index of the selected candidate.
        """
        raise NotImplementedError("Subclasses must implement this method")

    def judge_all(self, prompts: List[str], completions: List[List[str]], pbar: bool = True) -> List[int]:
        """
        Executes the judge function in parallel.

        Args:
            prompts (List[str]): List of prompts.
            completions (List[List[str]]): List of completion pairs for each prompt.
            pbar (bool, optional): Whether to show a progress bar.

        Returns:
            List[int]: List of chosen completion indices.
        """
        total_tasks = len(completions) * len(completions[0])
        progress_bar = tqdm(total=total_tasks, desc=f"Judging with {self.__class__.__name__}", leave=False, disable=not pbar)

        def wrapped_judge(prompt, completion):
            result = self.judge(prompt, completion)
            progress_bar.update(1)
            return result

        with concurrent.futures.ThreadPoolExecutor() as executor:
            choices = list(executor.map(wrapped_judge, prompts, completions))

        progress_bar.close()
        return choices

    def run(self, prompts: List[str], completions: List[List[str]], display: bool = False, pbar: bool = True, **kwargs) -> List[int]:
        """
        Evaluates multiple prompts and their response candidates.

        Args:
            prompts (List[str]): List of prompts.
            completions (List[List[str]]): List of completion pairs for each prompt.
            display (bool, optional): Whether to log judgment results. Defaults to False.
            pbar (bool, optional): Whether to display progress bar. Defaults to True.

        Returns:
            List[int]: List of selected response indices.
        """
        choices = self.judge_all(prompts, completions, pbar=pbar, **kwargs)

        if display:
            self._display_results(prompts, completions, choices) # Not shown here, simply logs results

        return choices


class LengthPairwiseJudge(PairwiseJudge):
    """Judges based on response length objective"""

    def __init__(self, objective: Literal["shorter", "longer"], **kwargs):
        super().__init__(**kwargs)
        self.objective = objective

    def judge(self, prompt: str, candidates: List[str]) -> int:
        if self.objective == "shorter":
            return 0 if len(candidates[0]) <= len(candidates[1]) else 1
        elif self.objective == "longer":
            return 0 if len(candidates[0]) >= len(candidates[1]) else 1
```

## Online DPO Proposal (WIP)
Starting from the full DPO distributed recipe, create a new function which converts from a prompt batch into a preference batch. This function would live as part of the OnlineDPODistributed class, and would run in the training loop between loading a batch of prompts (using padded_collate) and outputting a batch of preference data (padded_collate_dpo).

So this prompt to preference function would involve:
1. Get prompt batch
2. Run two completions per batch using the policy model
3. Run the given PairwiseJudge run to get chosen and rejected
4. Convert to preference dataset format

**Comment by ebsmothers** [comment]
Thanks @sam-pi for sharing the RFC! The general design of the judges abstraction looks pretty reasonable to me. A couple comments:

1) Personally I would split up the parent and child classes a little differently. (Admittedly some of this is personal preference but) I'd lean towards just having PairwiseJudge as a protocol and leaving all implementation details for the child classes. We do this a lot in other parts of our codebase too (e.g. [our tokenizers](https://github.com/pytorch/torchtune/blob/cf0142bd2eb5eabe33a3d43eca4f07ec48a8749f/torchtune/modules/transforms/tokenizers/_utils.py#L14-L72)).
2) (Kind of an extension of (1)..) I wonder whether `judge_all` could be a utility outside of a class, e.g. it takes in `judge_fn` or something as an argument, then wraps in the executor. The advantage of this is that then the judges become even easier to write/understand.
3) The online DPO proposal sounds reasonable at a high level, but would also like to see some code snippet to get a feel for how the judge interacts with it. Do you envision this as its own training recipe? (Seems that it does change the dataloading and training loop in a nontrivial way)

**Comment by sam-pi** [comment]
Thanks @ebsmothers, good points. 

1 and 2 sound good! I think some other points to consider:

    a. Should there be a Judge base class? 
    b. Should there be a RewardModel base class? Or is this just semantics? TRL has both and RewardModel seems to be a locally running model vs judge can be API calls or rules.
    c. How this can then be integrated or adapted for any online method (e.g. online DPO, GRPO, PPO, etc)?

3. As an example, I think in the recent GRPO additions we could abstract so that `shaped_correctness_reward` is a Judge / RewardModel function and `batch_shaped_correctness_reward` effectively is a version of `judge_all`: https://github.com/pytorch/torchtune/blob/main/torchtune/dev/grpo/rewards.py

I think the benefit of this approach would be to enable easy creation and integration of new types of reward models / judges into the existing recipe. It would also allow for easier side-by-side comparison of online DPO vs GRPO vs PPO, for instance, using the same reward models. 

Here's a quick rough example in online DPO - first generate 2 completions per input prompt, then detokenize and format the prompts/completions, then use `judge.run` to get chosen and rejected.

```
# Get input prompt ids from batch (using a prompt dataset)
prompt_input_ids = prompt_batch["input_ids"].to(device)
...
# Generate two completions
for _ in range(2):
    completion_tokens, _ = generate(model=model, prompt=prompt_input_ids, ...)
    completion_tokens_list.append(completion_tokens[len(prompt_input_ids)])
...
# Format prompts and completions
prompts = tokenizer.decode(prompt_input_ids ...)
completions = [tokenizer.decode(cmpl ...) for cmpl in completion_tokens_list]

# Get judge choices (chosen and rejected)
choices = judge.run(prompts, completions, display=display_results, pbar=True)

# Reformat to preference data
...
chosen_input_ids = [completion_tokens_list[chosen_idx, i] for i, chosen_idx in enumerate(choices)]
rejected_input_ids = [completion_tokens_list[1-chosen_idx, i] for i, chosen_idx in enumerate(choices)]
...
concatenated_input_ids, concatenated_labels = padded_collate_dpo(...)
batch = (concatenated_input_ids, concatenated_labels)
```

I'd also love any feedback from @RedTachyon

**Comment by sam-pi** [comment]
As a new revision, I would suggest instead there could be a "BaseJudge" with different methods useful for different types of RL or data generation:

```
class BaseJudge:
    """
    Base class for judge.

    Provides a common interface for rewards, preferences, and ranking.
    """

    def __init__(self, device_rank: int = None, weight: float = 1.0):
        self.device_rank = device_rank
        self.weight = weight

    def preference(
        self, prompt: str, candidates: List[str], display: bool = False
    ) -> Tuple[int, int]:
        """
        Abstract method to be implemented by subclasses.

        Args:
            prompt (str): The input prompt.
            candidates (List[str]): List of candidate responses.
            display (bool, optional): Whether to display the results. Defaults to False.

        Returns:
            Tuple[int, int]: Index of the chosen and rejected candidates.
        """
        raise NotImplementedError("Subclasses must implement this method")

    def reward(
        self, prompt: str, candidates: List[str], display: bool = False
    ) -> Tuple[List[float], List[float]]:
        """
        Abstract method to be implemented by subclasses.

        Args:
            prompt (str): The input prompt.
            candidates (List[str]): List of candidate responses.
            display (bool, optional): Whether to display the results. Defaults to False.

        Returns:
            Tuple[List[float], List[float]]: List of rewards and successes.
        """
        raise NotImplementedError("Subclasses must implement this method")

    def ranking(
        self, prompt: str, candidates: List[str], display: bool = False
    ) -> List[int]:
        """
        Ranks the candidates based on the judge's preference.

        Args:
            prompt (str): The input prompt.
            candidates (List[str]): List of candidate responses.
            display (bool, optional): Whether to display the results. Defaults to False.

        Returns:
            List[int]: List of candidate indices in the order of preference.
        """
        raise NotImplementedError("Subclasses must implement this method")
```

**Comment by joecummings** [comment]
> As a new revision, I would suggest instead there could be a "BaseJudge" with different methods useful for different types of RL or data generation:
> 
> ```
> class BaseJudge:
>     """
>     Base class for judge.
> 
>     Provides a common interface for rewards, preferences, and ranking.
>     """
> 
>     def __init__(self, device_rank: int = None, weight: float = 1.0):
>         self.device_rank = device_rank
>         self.weight = weight
> 
>     def preference(
>         self, prompt: str, candidates: List[str], display: bool = False
>     ) -> Tuple[int, int]:
>         """
>         Abstract method to be implemented by subclasses.
> 
>         Args:
>             prompt (str): The input prompt.
>             candidates (List[str]): List of candidate responses.
>             display (bool, optional): Whether to display the results. Defaults to False.
> 
>         Returns:
>             Tuple[int, int]: Index of the chosen and rejected candidates.
>         """
>         raise NotImplementedError("Subclasses must implement this method")
> 
>     def reward(
>         self, prompt: str, candidates: List[str], display: bool = False
>     ) -> Tuple[List[float], List[float]]:
>         """
>         Abstract method to be implemented by subclasses.
> 
>         Args:
>             prompt (str): The input prompt.
>             candidates (List[str]): List of candidate responses.
>             display (bool, optional): Whether to display the results. Defaults to False.
> 
>         Returns:
>             Tuple[List[float], List[float]]: List of rewards and successes.
>         """
>         raise NotImplementedError("Subclasses must implement this method")
> 
>     def ranking(
>         self, prompt: str, candidates: List[str], display: bool = False
>     ) -> List[int]:
>         """
>         Ranks the candidates based on the judge's preference.
> 
>         Args:
>             prompt (str): The input prompt.
>             candidates (List[str]): List of candidate responses.
>             display (bool, optional): Whether to display the results. Defaults to False.
> 
>         Returns:
>             List[int]: List of candidate indices in the order of preference.
>         """
>         raise NotImplementedError("Subclasses must implement this method")
> ```

@SalmanMohammadi @felipemello1 How does this fit into the reward structure you're thinking of with GRPO? I know it's not the exact same but having a similar API would help decrease cognitive load and changes in the future.
